dropbox/client.go

```diff
@@ -0,0 +1,239 @@
+	a := &DropboxAdapter{
+		conf:        conf,
+		ctx:         context.Background(),
+		doStop:      utils.NewEvent(),
+		dedupe:      make(map[string]int64),
+		cursor:      "",
+		initialized: false,
+	}
+
+	var err error
+	a.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	a.httpClient = &http.Client{
+		Timeout: 30 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 10 * time.Second,
+			}).Dial,
+		},
+	}
+
+	a.chStopped = make(chan struct{})
+	a.wgSenders.Add(1)
+	go a.fetchEvents()
+
+	go func() {
+		a.wgSenders.Wait()
+		close(a.chStopped)
+	}()
+
+	return a, a.chStopped, nil
+}
+
+func (a *DropboxAdapter) Close() error {
+	a.conf.ClientOptions.DebugLog("closing")
+	a.doStop.Set()
+	a.wgSenders.Wait()
+	err1 := a.uspClient.Drain(1 * time.Minute)
+	_, err2 := a.uspClient.Close()
+	a.httpClient.CloseIdleConnections()
+
+	if err1 != nil {
+		return err1
+	}
+	return err2
+}
+
+func (a *DropboxAdapter) fetchEvents() {
+	defer a.wgSenders.Done()
+	defer a.conf.ClientOptions.DebugLog("Dropbox event collection stopping")
+
+	var lastEventTime string
+
+	if !a.initialized {
+		items, cursor, lastTime, err := a.makeOneRequest("", "")
+		a.initialized = true
+		if err == nil {
+			a.cursor = cursor
+			lastEventTime = lastTime
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err != nil {
+					a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+					a.doStop.Set()
+					return
+				}
+			}
+		}
+	}
+
+	for !a.doStop.WaitFor(30 * time.Second) {
+		items, newCursor, lastTime, _ := a.makeOneRequest(a.cursor, lastEventTime)
+		if newCursor != "" {
+			a.cursor = newCursor
+		}
+		if lastTime != "" {
+			lastEventTime = lastTime
+		}
+		if items == nil {
+			continue
+		}
+
+		for _, item := range items {
+			msg := &protocol.DataMessage{
+				JsonPayload: item,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
+			}
+			if err := a.uspClient.Ship(msg, 10*time.Second); err != nil {
+				if err == uspclient.ErrorBufferFull {
+					a.conf.ClientOptions.OnWarning("stream falling behind")
+					err = a.uspClient.Ship(msg, 1*time.Hour)
+				}
+				if err != nil {
+					a.conf.ClientOptions.OnError(fmt.Errorf("Ship(): %v", err))
+					a.doStop.Set()
+					return
+				}
+			}
+		}
+	}
+}
+
+func (a *DropboxAdapter) makeOneRequest(cursor, lastEventTime string) ([]utils.Dict, string, string, error) {
+	var allItems []utils.Dict
+
+	var reqUrl string
+	var bodyPayload []byte
+	var err error
+
+	// Add 1 second to lastEventTime to avoid sending the same event repeatedly
```

Comment:
Hmmm, I'm worried about getting multiple events in 1 second and some getting dropped because of the limit of 100, then we +1 and never fetch those events.

I'm also thinking maybe we can try to make the `get_events` a parameter so that we can support other listings in the future without changing the API hit for existing users.

I can take a look later today unless you want to do it.

Reply:
we can change the limit, i didn't mean to leave that in there 🤣

also, i had the 1s because i can't dedupe based on an event id because they don't have IDs, which blows my mind 

open to any suggestions